### PR TITLE
Implement `LongCurrencyFormatter` for Long Currency Formatting

### PR DESCRIPTION
**Pull request by younies**
**Task:**
In this PR, implement `LongCurrencyFormatter` which handles formatting currencies in long formats.

**Related issues:**
https://github.com/unicode-org/icu4x/issues/5353

### CODE REVIEW

File: components/experimental/src/dimension/currency/format.rs [hunk 1]
```diff
@@ -7,15 +7,16 @@ use fixed_decimal::FixedDecimal;
...
 use crate::dimension::currency::options::Width;
 use crate::dimension::provider::currency;
 use crate::dimension::provider::currency::CurrencyEssentialsV1;
 
+use crate::dimension::currency::CurrencyCode;
+
 pub struct FormattedCurrency<'l> {
     pub(crate) value: &'l FixedDecimal,
-    pub(crate) currency_code: CurrencyCode,
+    pub(crate) currency_code: &'l CurrencyCode,
```
Comment (robertbastian) [inline components/experimental/src/dimension/currency/format.rs:19]: `CurrencyCode` is `TinyAsciiStr<3>`, which is smaller than a reference. Why change this to borrow?

File: components/experimental/src/dimension/currency/long_format.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+use writeable::Writeable;
+
+use crate::dimension::provider::currency_patterns::{CurrencyPatternsDataV1, PatternCount};
+use crate::dimension::provider::extended_currency::{Count, CurrencyExtendedDataV1};
+
+use super::CurrencyCode;
+
+pub struct LongFormattedCurrency<'l> {
+    pub(crate) value: &'l FixedDecimal,
+    pub(crate) currency_code: CurrencyCode,
```
Comment (robertbastian) [inline components/experimental/src/dimension/currency/long_format.rs:21]: here you have it by value, then why did you change the non-long one?

File: components/experimental/src/dimension/currency/long_format.rs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            .get(&count)
+            .or_else(|| display_names.get(&Count::Other))
+            // TOOD: check the cases of using `Count::DisplayName`
+            .unwrap_or_else(|| self.currency_code.0.as_str());
+
+        let patterns = &self.patterns.unit_patterns;
+        let pattern = patterns
+            .get(&pattern_count)
+            .or_else(|| patterns.get(&PatternCount::Other))
+            .ok_or(core::fmt::Error)?;
```
Comment (robertbastian) [inline components/experimental/src/dimension/currency/long_format.rs:51]: don't ever produce `core::fmt::Error`: https://doc.rust-lang.org/std/fmt/index.html#:~:text=Additionally%2C%20the%20return,up%20the%20stack.

File: components/experimental/src/dimension/currency/long_format.rs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+        // Parse the pattern string into a DoublePlaceholderPattern
+        // The pattern is expected to be in the form of "{0} {1}"
+        let pattern = DoublePlaceholderPattern::from_str(pattern).map_err(|_| core::fmt::Error)?;
```
Comment (robertbastian) [inline components/experimental/src/dimension/currency/long_format.rs:55]: I don't like adding this runtime parsing code, you should really change the data structs to use `Pattern` soon.

File: components/experimental/src/dimension/currency/long_format.rs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+        pattern
+            .interpolate((
+                self.fixed_decimal_formatter.format(self.value),
+                display_name,
+            ))
+            .write_to(sink)?;
+
+        Ok(())
```
Comment (robertbastian) [inline components/experimental/src/dimension/currency/long_format.rs:64]:
```suggestion
            .write_to(sink)
```

File: components/experimental/src/dimension/provider/currency_patterns.rs [hunk 6]
```diff
@@ -52,3 +53,17 @@ pub enum PatternCount {
     /// UnitPattern `other`.
     Other = 5,
 }
+
+impl From<PluralCategory> for PatternCount {
```
Comment (robertbastian) [inline components/experimental/src/dimension/provider/currency_patterns.rs:57]: why do I keep seeing conversions between `Count` enums? You're adding two (more?) here, is that necessary?

File: components/experimental/src/dimension/currency/format.rs (hunk 1 above)
Comment (younies) [reply components/experimental/src/dimension/currency/format.rs]: you are right, even the architecture is 32-bit still less or the same.

File: components/experimental/src/dimension/currency/formatter.rs [hunk 7]
```diff
@@ -121,15 +116,15 @@ impl CurrencyFormatter {
...
+    /// let formatted_currency = fmt.format_fixed_decimal(&value, &currency_code);
     /// let mut sink = String::new();
     /// formatted_currency.write_to(&mut sink).unwrap();
     /// assert_eq!(sink.as_str(), "$12,345.67");
     /// ```
     pub fn format_fixed_decimal<'l>(
         &'l self,
         value: &'l FixedDecimal,
-        currency_code: CurrencyCode,
+        currency_code: &'l CurrencyCode,
```
[the remark this answers is not in the document]
Comment (younies) [reply components/experimental/src/dimension/currency/formatter.rs]: done

File: components/experimental/src/dimension/currency/long_format.rs (hunk 3 above)
Comment (younies) [reply components/experimental/src/dimension/currency/long_format.rs:51]: mmmmm

in this case we have two options,

1. to not return error at all and use default pattern
2. move this logic to `format_fixed_decimal` method 

I prefer option 2.

File: components/experimental/src/dimension/currency/long_format.rs (hunk 4 above)
Comment (younies) [reply components/experimental/src/dimension/currency/long_format.rs:55]: yes, we need that of course.

File: components/experimental/src/dimension/provider/currency_patterns.rs (hunk 6 above)
Comment (younies) [reply components/experimental/src/dimension/provider/currency_patterns.rs:57]: Yes, one from `PatternCount` and the other for `Count` for currency, they are different counts with different fields.

This something we need to generalize across ICU4X soon

**Review by robertbastian — COMMENTED** [review]
.

File: components/experimental/src/dimension/currency/long_format.rs (hunk 4 above)
Comment (younies) [reply components/experimental/src/dimension/currency/long_format.rs]: done

File: components/experimental/src/dimension/provider/currency_patterns.rs (hunk 6 above)
Comment (robertbastian) [reply components/experimental/src/dimension/provider/currency_patterns.rs:57]: This one seems to be identical, there are no extra fields

File: components/experimental/src/dimension/currency/long_format.rs (hunk 3 above)
Comment (younies) [reply components/experimental/src/dimension/currency/long_format.rs]: Actually, `other` must be present. I added this condition to the datagen`

File: components/experimental/src/dimension/provider/currency_patterns.rs (hunk 6 above)
Comment (younies) [reply components/experimental/src/dimension/provider/currency_patterns.rs:57]: https://github.com/unicode-org/icu4x/issues/5369

File: components/experimental/src/dimension/currency/long_format.rs (hunk 3 above)
Comment (younies) [reply components/experimental/src/dimension/currency/long_format.rs]: https://github.com/unicode-org/icu4x/issues/5374

File: provider/data/experimental/fingerprints.csv [hunk 8]
```diff
@@ -29347,15 +29347,14 @@ currency/extended@1, zu/YER, 68B, 52c541c97403eaa6
...
-currency/patterns@1, <total>, 617B, 9 unique payloads
+currency/patterns@1, <lookup>, 91B, 13 identifiers
+currency/patterns@1, <total>, 569B, 8 unique payloads
 currency/patterns@1, blo, 82B, da128faf13fad0d5
 currency/patterns@1, ceb, 72B, db7c7c84fecaf3d0
 currency/patterns@1, es-GT, 72B, -> ceb
 currency/patterns@1, ja, 61B, 9af35aad6ebc7d70
 currency/patterns@1, my, 62B, 86cee97085536d8b
 currency/patterns@1, ro, 85B, 629b2b7dcea97875
-currency/patterns@1, sd, 48B, 3a6d9523170345d0
```
Comment (robertbastian) [inline provider/data/experimental/fingerprints.csv:29358]: this seems to be a CLDR bug (missing `Other`), can you file an issue?

File: components/experimental/src/dimension/currency/long_format.rs (hunk 3 above)
Comment (robertbastian) [reply components/experimental/src/dimension/currency/long_format.rs]: I'd prefer if this was represented in the type system, instead of with a trust-me-unwrap.

File: provider/data/experimental/fingerprints.csv (hunk 8 above)
Comment (younies) [reply provider/data/experimental/fingerprints.csv:29358]: I have already filed this issue: https://github.com/unicode-org/icu4x/issues/5374. I want to discuss it with @sffc today to ensure that it is a CLDR issue before filing it. If it is a CLDR issue, we need to decide what to do in the meantime. We can use the default `{0} {1}` or the pattern from `en-US`.

File: components/experimental/src/dimension/provider/currency_patterns.rs (hunk 6 above)
Comment (younies) [reply components/experimental/src/dimension/provider/currency_patterns.rs:57]: https://github.com/unicode-org/icu4x/pull/5373

File: components/experimental/src/dimension/currency/long_format.rs (hunk 3 above)
Comment (younies) [reply components/experimental/src/dimension/currency/long_format.rs]: good point: https://github.com/unicode-org/icu4x/issues/5374#issuecomment-2291303219

**Review by robertbastian — COMMENTED** [review]
.

File: components/experimental/src/dimension/currency/long_format.rs (hunk 3 above)
Comment (younies) [reply components/experimental/src/dimension/currency/long_format.rs]: done :)

File: components/experimental/src/dimension/provider/currency_patterns.rs (hunk 6 above)
Comment (younies) [reply components/experimental/src/dimension/provider/currency_patterns.rs]: done

File: provider/data/experimental/fingerprints.csv (hunk 8 above)
Comment (younies) [reply provider/data/experimental/fingerprints.csv]: done

**Review by kartva — APPROVED** [review]
Testing approving

**Review by sffc — APPROVED** [review]
